crates/task-center/src/lib.rs

```diff
@@ -0,0 +1,554 @@
+    #[instrument(level = "error", skip(self, exit_code))]
+    pub async fn shutdown_node(&self, reason: &str, exit_code: i32) {
+        let inner = self.inner.clone();
+        if inner
+            .shutdown_requested
+            .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
+            .unwrap_or_else(|e| e)
+        {
+            // already shutting down....
+            return;
+        }
+        let start = Instant::now();
+        inner.current_exit_code.store(exit_code, Ordering::Release);
+
+        if exit_code != 0 {
+            info!("** Shutdown requested");
+        } else {
+            warn!("** Shutdown requested");
+        }
+        // todo: define a controlled shutdown by task kind as needed.
+        self.cancel_tasks(None, None).await;
+        // notify outer components that we have completed the shutdown.
+        self.inner.global_cancel_token.cancel();
+        info!("** Shutdown completed in {:?}", start.elapsed());
+    }
+
+    #[track_caller]
+    fn spawn_inner<F>(
+        &self,
+        kind: TaskKind,
+        name: &'static str,
+        partition_id: Option<PartitionId>,
+        shutdown_node_on_failure: bool,
+        cancel: CancellationToken,
+        future: F,
+    ) -> TaskId
+    where
+        F: Future<Output = anyhow::Result<()>> + Send + 'static,
+    {
+        let inner = self.inner.clone();
+        let id = TaskId::from(NEXT_TASK_ID.fetch_add(1, Ordering::SeqCst));
+        let task = Arc::new(Task {
+            id,
+            name,
+            kind,
+            partition_id,
+            cancel: cancel.clone(),
+            join_handle: Mutex::new(None),
+        });
+
+        inner.tasks.lock().unwrap().insert(id, Arc::clone(&task));
+
+        let mut handle_mut = task.join_handle.lock().unwrap();
+
+        let task_cloned = Arc::clone(&task);
+        let join_handle = inner.runtime.spawn(wrapper(
+            self.clone(),
+            id,
+            kind,
+            task_cloned,
+            cancel,
+            shutdown_node_on_failure,
+            future,
+        ));
+        *handle_mut = Some(join_handle);
+        drop(handle_mut);
+
+        // Task is ready
+        id
+    }
+
+    /// Launch a new task
+    /// Note: if shutdown_node_on_failure is set to true, failure of the task
+    /// will request a shutdown of the task center.
+    #[track_caller]
+    pub fn spawn<F>(
+        &self,
+        kind: TaskKind,
+        name: &'static str,
+        partition_id: Option<PartitionId>,
+        shutdown_node_on_failure: bool,
+        future: F,
+    ) -> Result<TaskId, ShutdownError>
+    where
+        F: Future<Output = anyhow::Result<()>> + Send + 'static,
+    {
+        let inner = self.inner.clone();
+        if inner.shutdown_requested.load(Ordering::Acquire) {
+            return Err(ShutdownError);
+        }
+        Ok(self.spawn_unchecked(kind, name, partition_id, shutdown_node_on_failure, future))
```

Comment:
If we check after `spawn_unchecked` whether `inner.shutdown_requested` is true, then we could also stop those tasks that were started after a potential shutdown.